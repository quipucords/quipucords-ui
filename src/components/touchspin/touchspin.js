import React from 'react';
import PropTypes from 'prop-types';
import { InputGroup, Button, TextInput, ButtonVariant } from '@patternfly/react-core';
import { PlusIcon, MinusIcon } from '@patternfly/react-icons';
import helpers from '../../common/helpers';

// FixMe: remove the eslint-disable
class TouchSpin extends React.Component {
  constructor(props) {
    super(props);

    // eslint-disable-next-line
    this.state = {
      displayValue: Number.parseInt(props.value, 10),
      maxValue: props.maxValue,
      minValue: props.minValue
    };

    this.timer = null;
  }

  normalizeBetween = (value, min, max) => {
    if (min !== undefined && max !== undefined) {
      return Math.max(Math.min(value, max), min);
    }
    if (value <= min) {
      return min;
    }
    if (value >= max) {
      return max;
    }
    return value;
  };

  onFocusMax = () => {
    const { maxValue, displayValue } = this.state;

    const setTimer = (time = 500) => {
      clearTimeout(this.timer);

      this.timer = setTimeout(() => {
        this.onPlus();

        if (displayValue < maxValue) {
          setTimer(0);
        }
      }, time);
    };

    setTimer();
  };

  onFocusMin = () => {
    const { minValue, displayValue } = this.state;

    const setTimer = (time = 500) => {
      clearTimeout(this.timer);

      this.timer = setTimeout(() => {
        this.onMinus();

        if (displayValue > minValue) {
          setTimer(0);
        }
      }, time);
    };

    setTimer();
  };

  onBlur = () => {
    clearTimeout(this.timer);
  };

  onMinus = () => {
    const { displayValue, minValue, maxValue } = this.state;

    const updatedValue = this.normalizeBetween(displayValue - 1, minValue, maxValue);
    this.onUpdateValue({ target: { value: updatedValue } });
  };

  onPlus = () => {
    const { displayValue, minValue, maxValue } = this.state;

    const updatedValue = this.normalizeBetween(displayValue + 1, minValue, maxValue);
    this.onUpdateValue({ target: { value: updatedValue } });
  };

  onUpdateValue = event => {
    const { maxValue, minValue } = this.state;
    const { name, onChange } = this.props;

    let parsedValue = Number.parseInt(event.target.value, 10);
    parsedValue = Number.isNaN(parsedValue) ? minValue : parsedValue;

    if (parsedValue > maxValue) {
      parsedValue = maxValue;
    }

    if (parsedValue < minValue) {
      parsedValue = minValue;
    }

    clearTimeout(this.timer);

    this.setState({
      displayValue: parsedValue
    });

    const mockTarget = {
      ...event.target,
      name,
      value: parsedValue
    };

    const mockEvent = {
      ...mockTarget,
      target: { ...mockTarget },
      currentTarget: { ...mockTarget },
      persist: helpers.noop
    };

    return onChange(mockEvent);
  };

  render() {
    const { displayValue, maxValue, minValue } = this.state;
    const { className, labelMax, labelMaxDescription, labelMin, labelMinDescription, name } = this.props;

    return (
      <div className={`pf-c-number-input ${className}`}>
        <InputGroup>
          <Button
            onClick={this.onMinus}
            variant={ButtonVariant.control}
            onMouseDown={this.onFocusMin}
            onMouseUp={this.onBlur}
            onMouseLeave={this.onBlur}
            isDisabled={displayValue <= minValue}
            aria-label={labelMinDescription}
          >
            {labelMin}
          </Button>
          <TextInput
            type="number"
            name={name}
            value={displayValue}
            onChange={(_, event) => this.onUpdateValue(event)}
            aria-label="Number Input"
          />
          <Button
            variant={ButtonVariant.control}
            onClick={this.onPlus}
            onMouseDown={this.onFocusMax}
            onMouseUp={this.onBlur}
            onMouseLeave={this.onBlur}
            isDisabled={displayValue >= maxValue}
            aria-label={labelMaxDescription}
          >
            {labelMax}
          </Button>
        </InputGroup>
      </div>
    );
  }
}

TouchSpin.propTypes = {
  className: PropTypes.string,
  labelMax: PropTypes.node,
  labelMaxDescription: PropTypes.string,
  labelMin: PropTypes.node,
  labelMinDescription: PropTypes.string,
  maxValue: PropTypes.number,
  minValue: PropTypes.number,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func,
  value: PropTypes.number
};

TouchSpin.defaultProps = {
  className: '',
  labelMax: <PlusIcon />,
  labelMaxDescription: 'Increase number button',
  labelMin: <MinusIcon />,
  labelMinDescription: 'Decrease number button',
  maxValue: 100,
  minValue: 0,
  onChange: helpers.noop,
  value: 50
};

export { TouchSpin as default, TouchSpin };
