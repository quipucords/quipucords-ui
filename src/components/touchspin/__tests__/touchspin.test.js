import React from 'react';
import { mount } from 'enzyme';
import { TextInput } from '@patternfly/react-core';
import { TouchSpin } from '../touchspin';

describe('TouchSpin Component', () => {
  it('should render a basic component', () => {
    const props = {
      name: 'lorem'
    };

    const component = mount(<TouchSpin {...props} />);

    const componentInstance = component.instance();
    expect(componentInstance.props).toMatchSnapshot('initial props');
    expect(componentInstance.state).toMatchSnapshot('initial state');

    expect(component.render()).toMatchSnapshot('basic render');
  });

  it('should handle onchange events', () => {
    const props = {
      name: 'lorem',
      value: 1
    };

    const component = mount(<TouchSpin {...props} />);

    component.find('button[aria-label="Increase number button"]').simulate('click');
    expect(component.find(TextInput).props()?.value).toBe(2);

    component.find('button[aria-label="Decrease number button"]').simulate('click');
    expect(component.find(TextInput).props()?.value).toBe(1);
  });
});
