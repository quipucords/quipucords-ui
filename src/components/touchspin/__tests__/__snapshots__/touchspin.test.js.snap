// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TouchSpin Component should render a basic component: basic render 1`] = `
<div
  class="pf-c-number-input "
>
  <div
    class="pf-c-input-group"
  >
    <button
      aria-disabled="false"
      aria-label="Decrease number button"
      class="pf-c-button pf-m-control"
      data-ouia-component-id="OUIA-Generated-Button-control-1"
      data-ouia-component-type="PF4/Button"
      data-ouia-safe="true"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 448 512"
        width="1em"
      >
        <path
          d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
        />
      </svg>
    </button>
    <input
      aria-invalid="false"
      aria-label="Number Input"
      class="pf-c-form-control"
      data-ouia-component-id="OUIA-Generated-TextInputBase-1"
      data-ouia-component-type="PF4/TextInput"
      data-ouia-safe="true"
      name="lorem"
      type="number"
      value="50"
    />
    <button
      aria-disabled="false"
      aria-label="Increase number button"
      class="pf-c-button pf-m-control"
      data-ouia-component-id="OUIA-Generated-Button-control-2"
      data-ouia-component-type="PF4/Button"
      data-ouia-safe="true"
      type="button"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        role="img"
        style="vertical-align: -0.125em;"
        viewBox="0 0 448 512"
        width="1em"
      >
        <path
          d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
        />
      </svg>
    </button>
  </div>
</div>
`;

exports[`TouchSpin Component should render a basic component: initial props 1`] = `
Object {
  "className": "",
  "labelMax": <PlusIcon
    color="currentColor"
    noVerticalAlign={false}
    size="sm"
  />,
  "labelMaxDescription": "Increase number button",
  "labelMin": <MinusIcon
    color="currentColor"
    noVerticalAlign={false}
    size="sm"
  />,
  "labelMinDescription": "Decrease number button",
  "maxValue": 100,
  "minValue": 0,
  "name": "lorem",
  "onChange": [Function],
  "value": 50,
}
`;

exports[`TouchSpin Component should render a basic component: initial state 1`] = `
Object {
  "displayValue": 50,
  "maxValue": 100,
  "minValue": 0,
}
`;
